### ISSUE TITLE
Support string interpolation in build files

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#1464]
The current string formatting can be very cumbersome/difficult to read. We should support [Python 3.6's string interpolation](https://www.python.org/dev/peps/pep-0498/):

```meson
bar = 'something'
value = f'foo/{bar}/baz'
# value is now 'foo/something/baz'
nosubst = 'foo/{bar}/baz'
# nosubst is 'foo/{bar}/baz'
```

This is a very concise, explicit, and convenient way of substituting variables into strings. If a `bar` variable for a `{bar}` entry in a string is not found, we should error out.

We have an additional use-case besides variables. Options. Substituting project options would work like this:

```meson
executale('foo', 'prog.c',
  install : true,
  install_dir : o'{prefix}/{libexecdir}')
```

i.e., `o''` instead of `f''`.

**Comment by TingPing** [comment]
> install_dir : o'{prefix}/{libexecdir}')

That example fails to handle what `join_paths()` does which makes it ugly again.

**Comment by nirbheek** [comment]
Then we should make it handle that! There is precedent for that in pathlib.

**Comment by jpakkane** [comment]
Let's not do magic substitution variables if we can help it. This can be done with a new method in the string class, something like this:

    var = 'x'
    foo = 'foo/@var@/boo'.autoformat() # yields foo/x/bar

And then some sort of syntax to get values of options.

**Comment by nirbheek** [comment]
I don't think this is magic in any way since it's opt-in and is well-specified via the PEP. It's basically `f('str')` → `f'str'`. I think there is real value in conciseness (that is clear and explicit) that is often overlooked in language design.

**Comment by rulatir** [comment]
Irrelevant. Meson language is not human-writable, and you **will** end up generating it anyway.

**Comment by eli-schwartz** [comment]
meson can reasonably define the `/` operator to join strings by treating them as paths, but it cannot know if `'/'` is intended to be a string or a path. Furthermore, I don't see how this is better than 

```
prefix = get_option('prefix')
libexecdir = get_option('libexecdir')

[100 lines of intervening meson code]

executable('foo', 'prog.c',
  install : true,
  install_dir : prefix / libexecdir)
```

It's a confusing example anyway since install_dir automatically figures out what a relative path means without joining it to prefix...

**Comment by eli-schwartz** [comment]
This was implemented in #8485 -- only simple types are supported, so `f''` strings replace string/numeric/boolean types, but there are no `o''` strings to format options.

Also it uses the meson-standard at symbol, rather than curly brackets.
